File: inputFiles/Files_7Feb2016/removeCorrupted.py
# def parseInputFileList (fileName) :
#     filelist = []
#     with open (fileName) as fIn:
#         for line in fIn:
#             line = (line.split("#")[0]).strip()
#             if line:
#                 filelist.append(line)
# return filelist

import argparse
import ROOT 
import os
import sys

parser = argparse.ArgumentParser(description='Command line parser of plotting options')
parser.add_argument('--filelist', dest='filelist', help='input file list', default=None)
args = parser.parse_args()

# zombies = []
# recover = []
corr = []
good = []
# totfiles = 0
infile = open(args.filelist)
for idx , line in enumerate(infile):
    line = (line.split("#")[0]).strip()
    if idx % 1000 == 0: print " >> file num" , idx
    if line:
        # totfiles += 1
        rfile = ROOT.TFile.Open(line)
        if not rfile:
            corr.append(line)
        elif rfile.IsZombie():
            corr.append(line)
        elif rfile.TestBit(ROOT.TFile.kRecovered):
            corr.append(line)
        else:
            good.append(line)
        
        # close only if it was opened
        if rfile:
            rfile.Close()
infile.close()

if len(corr) == 0:
    print "... all ok for", args.filelist
    sys.exit()

print "... ** I found" , len(corr) , "files (", 1.*len(corr)/(len(good)+len(corr)), ") corrupted, making new list"
newname = args.filelist + '_WITHCORRUPTED'

# print good
# print corr
# print newname

os.system('mv %s %s' % (args.filelist, newname))
fnew = open (args.filelist, 'w')
for l in good:
    fnew.write(l+'\n')
fnew.close()

File: inputFiles/Files_7Feb2016/removeCorrupted.cpp
// NOTE: I tried to write this macro in C, but it is as fast as the python one (limiting factor ids xrootd opening)
// so let's use the python one that is nicer

#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include "TFile.h"
#include <stdlib.h>
using namespace std;
int main(int argc, char** argv)
{   
    if (argc < 2)
    {
        cout << ".. no input list set" << endl;
        return 1;
    }

    string infilename = argv[1];

    vector<string> corr;
    vector<string> good;

    cout << "=== opening file === " << infilename << endl;
    std::ifstream infile(infilename);
    std::string line;
    int idx = 0;
    while (std::getline(infile, line))
    {
        if (idx % 1000 == 0)
            cout << "file num: " << idx << endl;

        line = line.substr(0, line.find("#", 0)); // remove comments introduced by #
        //cout << line << endl;
        while (line.find(" ") != std::string::npos) line = line.erase(line.find(" "), 1); // remove white spaces
        while (line.find("\n") != std::string::npos) line = line.erase(line.find("\n"), 1); // remove new line characters
        while (line.find("\r") != std::string::npos) line = line.erase(line.find("\r"), 1); // remove carriage return characters
        if (!line.empty()) // skip empty lines
        {
            idx += 1;
            TFile* rfile = TFile::Open(line.c_str());
            
            if (!rfile)
                corr.push_back(line);
            else if (rfile->IsZombie())
                corr.push_back(line);
            else if (rfile->TestBit(TFile::kRecovered))
                corr.push_back(line);
            else
                good.push_back(line);

            if (rfile)
                rfile->Close();
        
            // delete rfile
        }
    }

    infile.close();

    if (corr.size() == 0)
    {
        cout << "... all ok for" << infilename << endl;
        return 0;
    }

    cout << "... ** I found " << corr.size() << "files (" << 1.*corr.size()/(good.size()+corr.size()) << ") corrupted, making new list" << endl;

    string newname = (infilename + string("_WITHCORRUPTED"));

    string command = "mv ";
    command += infilename;
    command += " ";
    command += newname;
    system(command.c_str());

    std::ofstream ofile(infilename);
    for (string x : good)
        ofile << x << endl;
    ofile.close();
}
